### PR TITLE
KAFKA-12648: handle MissingSourceTopicException for named topologies

### PR DESCRIPTION
**Pull request by ableegoldman**
Another source of flakiness we found in the NamedTopologyIntegrationTest was an ocasional MissingSourceTopicException that was causing the application to shut down. We created all source topics ahead of time in the tests, leading us to discover this [race condition](https://issues.apache.org/jira/browse/KAFKA-13543) in the consumer client which can lead to spurious MissingSourceTopicExceptions when the metadata hasn't finished updating after a change in the consumer's subscription.

In addition to finding a workaround for this bug, throwing this MissingSourceTopicException and shutting down the entire app is itself a bug in the NamedTopology feature -- we should not stop all clients and prevent any further processing of the completely valid topologies just because one (or more) topologies were added that are missing their source topics. We can just remove those topologies from the assignment for the time being, and wait until the metadata has finished updating or the user has created the input topics to start assigning tasks from them.

This work entails two things:
a) Avoid throwing a MissingSourceTopicException inside the #assign method when named topologies are used, and just remove those topologies which are missing any of their input topics from the assignment. 
b) Trigger the uncaught exception handler with a MissingSourceTopicException for each of the topologies that are missing topics, but don't shut down the thread -- we just want to make sure this issue is made visible to the user.

This PR addresses (a), while (b) is done in the followup PR https://github.com/apache/kafka/pull/11686

### CODE REVIEW

File: checkstyle/suppressions.xml [hunk 1]
```diff
@@ -158,7 +158,7 @@
               files="(KafkaStreams|KStreamImpl|KTableImpl|InternalTopologyBuilder|StreamsPartitionAssignor|StreamThread).java"/>
 
     <suppress checks="MethodLength"
-              files="KTableImpl.java"/>
+              files="(KTableImpl|StreamsPartitionAssignor).java"/>
```
Comment (cadonna) [inline checkstyle/suppressions.xml]: Sorry for the standard question: Is this really necessary?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RepartitionTopics.java [hunk 2]
```diff
@@ -85,24 +86,51 @@ public void setup() {
...
+            if (missingSourceTopicsPerTopology.isEmpty()) {
+                allRepartitionTopicConfigs.putAll(repartitionTopicConfigsPerTopology);
+                allTopicsInfo.addAll(topology.getValue());
+            } else {
+                topologiesWithMissingSourceTopics.put(topologyName, missingSourceTopicsPerTopology);
+                log.error("Topology {} was missing source topics {} and will be excluded from the current assignment, "
+                              + "this can be due to the consumer client's metadata being stale or because they have "
+                              + "not been created yet. Please verify that you have created all input topics. When the "
+                              + "metadata is updated a new rebalance will be kicked off automatically and the topology "
+                              + "will retried at that time.", topologyName, missingSourceTopicsPerTopology);
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RepartitionTopics.java]:
```suggestion
                              + "will be retried at that time.", topologyName, missingSourceTopicsPerTopology);
```

Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RepartitionTopics.java]: Should we be clearer here that if users created the input topics and this error occurs because of client's metadata being stale, they actually do not need to do anything except wait, while if they did not create the input topics they should create them? At least that is how I understand it.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 3]
```diff
@@ -1137,6 +1138,8 @@ private void completeShutdown(final boolean cleanRun) {
 
         log.info("Shutting down");
 
+        topologyMetadata.unregisterThread(threadMetadata.threadName());
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:1141]: Why this move? Just curious...

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RepartitionTopics.java [hunk 4]
```diff
@@ -85,24 +86,51 @@ public void setup() {
...
-            repartitionTopicConfigs.putAll(topicsInfo.repartitionSourceTopics.values().stream()
-                .collect(Collectors.toMap(InternalTopicConfig::name, topicConfig -> topicConfig)));
+    /**
+     * @param topicGroups                            information about the topic groups (subtopologies) in this application
+     * @param clusterMetadata                        cluster metadata, eg which topics exist on the brokers
+     * @param topologiesWithMissingSourceTopics  set of missing user input topics, to be filled in by this method
+     */
+    private Map<String, InternalTopicConfig> computeRepartitionTopicConfig(final Map<String, Collection<TopicsInfo>> topicGroups,
+                                                                           final Cluster clusterMetadata,
+                                                                           final Map<String, Set<String>> topologiesWithMissingSourceTopics) {
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RepartitionTopics.java]: nit: Could you rename `topologiesWithMissingSourceTopics` to `missingExternalSourceTopicsPerTopology` or the other way around. It makes reading the code a bit easier.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java [hunk 5]
```diff
@@ -380,15 +380,37 @@ public GroupAssignment assign(final Cluster metadata, final GroupSubscription gr
...
+            );
+
+            final Map<String, Set<String>> missingExternalSourceTopicsPerTopology = repartitionTopics.setup();
+            if (!missingExternalSourceTopicsPerTopology.isEmpty()) {
+                log.error("The following source topics are missing/unknown: {}. Please make sure all source topics " +
+                              "have been pre-created before starting the Streams application. ",
+                          taskManager.topologyMetadata().hasNamedTopologies() ?
+                              missingExternalSourceTopicsPerTopology :
+                              missingExternalSourceTopicsPerTopology.values()
+                );
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: Why do you log again the same error with other words? Might be confusing during debugging and bloats the logs.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java [hunk 6]
```diff
@@ -380,15 +380,37 @@ public GroupAssignment assign(final Cluster metadata, final GroupSubscription gr
...
+                if (!taskManager.topologyMetadata().hasNamedTopologies()) {
+                    throw new MissingSourceTopicException("Missing source topics.");
+                }
+            }
+
+            final Map<TopicPartition, PartitionInfo> allRepartitionTopicPartitions = repartitionTopics.topicPartitionsInfo();
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: Could you put this block again into a method and remove the checkstyle suppression?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java [hunk 7]
```diff
@@ -380,15 +380,37 @@ public GroupAssignment assign(final Cluster metadata, final GroupSubscription gr
...
+            if (!missingExternalSourceTopicsPerTopology.isEmpty()) {
+                log.error("The following source topics are missing/unknown: {}. Please make sure all source topics " +
+                              "have been pre-created before starting the Streams application. ",
+                          taskManager.topologyMetadata().hasNamedTopologies() ?
+                              missingExternalSourceTopicsPerTopology :
+                              missingExternalSourceTopicsPerTopology.values()
+                );
+                if (!taskManager.topologyMetadata().hasNamedTopologies()) {
+                    throw new MissingSourceTopicException("Missing source topics.");
+                }
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: What happens if all named topologies have missing source topics? Do you account for that somewhere else?

**Review by cadonna — COMMENTED** [review]
Hi @ableegoldman !

Thanks for the PR!

Here my feedback!

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java [hunk 8]
```diff
@@ -436,9 +436,25 @@ public String getStoreForChangelogTopic(final String topicName) {
         return sourceTopics;
     }
 
-    public Map<Subtopology, TopicsInfo> topicGroups() {
+    /**
+     * @param topologiesToExclude the names of any topologies to exclude from the returned topic groups,
+     *                            eg because they have missing source topics and can't be processed yet
+     */
+    public Map<Subtopology, TopicsInfo> topicGroups(final Set<String> topologiesToExclude) {
```
Comment (wcarlson5) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java:447]: nit: can we update the name to be more descriptive?

**Review by wcarlson5 — COMMENTED** [review]
I don't have much to add to @cadonna's comments but its looking good so far

**Comment by guozhangwang** [comment]
@ableegoldman I checked the source code trying to reason what's the difference between 1) subscribe for the first time, and then poll to potentially rebalance, 2) re-subscribe, and then poll to potentially rebalance.

For 1) we have the logic to enforce metadata being fresh until we call `ensureActiveGroupensureActiveGroup`, so that's cleared as long as all members `subscribe` with the same topics.

For 2) the same logic is there, so the member kicking start the rebalance with the new join-group should always have the updated metadata, it's just that the leader who's actually triggering assign may not get the `subscribe` call yet --- since if the leader has triggered `subscribe`, it will set the `needPartialUpdate` flag in metadata and upon being requested to re-join, it would still wait until metadata to be refreshed.

So that case in 2) would be similar to 1) where you have two members starting up, one that called `subscribe(A,B)` while the other one called `subscribe(A)` but the second one is picked as the leader. So in Streams' scenario, it's similar to two instances with different topologies but configured with the same `group id` --- in the past we treat it as fatal since instances cannot change their topology after starting up, but now they can, and we do not know if this is just temporary that soon the instances would change their topologies to match, or never --- so I think we should now treat it as transient. But the question is then, do we still need part1 of the PR if we do part2 to not treat it as final? If we just expose the `MissingSourceTopicException` to the handler and explain that "this may be transient, but we do not know" but now killing the thread, would that be sufficient?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java [hunk 9]
```diff
@@ -317,7 +317,7 @@ public GroupAssignment assign(final Cluster metadata, final GroupSubscription gr
         int minSupportedMetadataVersion = LATEST_SUPPORTED_VERSION;
 
         boolean shutdownRequested = false;
-        boolean assignementErrorFound = false;
+        boolean assignmentErrorFound = false;
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:320]: Ah good find :)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java [hunk 10]
```diff
@@ -380,15 +380,37 @@ public GroupAssignment assign(final Cluster metadata, final GroupSubscription gr
             // parse the topology to determine the repartition source topics,
             // making sure they are created with the number of partitions as
             // the maximum of the depending sub-topologies source topics' number of partitions
-            final Map<TopicPartition, PartitionInfo> allRepartitionTopicPartitions = prepareRepartitionTopics(metadata);
+            final RepartitionTopics repartitionTopics = new RepartitionTopics(
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: Why inline this function? Maybe we can still avoid `StreamsPartitionAssignor` to exclude the function length code if we extra enough LOC?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java (hunk 7 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: I think that's covered in the stream-thread where we would poll much longer when all tasks have missing topics.

**Review by guozhangwang — COMMENTED** [review]
LGTM overall.

Personally I'd prefer to refactor the `setup` call where we first trim the topology first from the metadata, and maybe at the sub-topology granularity. Of course for now as we create internal topics here all sub-topologies of the same topology would either be excluded all together or not, but in the future maybe some sub-topologies can be instantiated while others of the same topology will not.

The leader would then get the "partial" sub-topologies and then do the assignment, together with the list of trimmed sub-topologies, which we can decide notifying the users and move on. I felt this is a cleaner workflow and with finer "partial task assignment" granularity.

I know that's a bigger refactoring, so I think we can do that latter than in this PR.

**Comment by cadonna** [comment]
@ableegoldman There are checkstyle errors in the builds.

File: checkstyle/suppressions.xml (hunk 1 above)
Comment (ableegoldman) [reply checkstyle/suppressions.xml]: I would prefer to leave out any nontrivial refactoring such as breaking up the assignor's ridiculously long `assign` method, though I do agree that it needs to be broken up (again -- I actually did a pretty hefty refactoring to clean this method up a while back, I guess it's just gotten out of control again since then :/ )

edit: nevermind, I will clean it up in this PR and remove the suppression. Lol it's hard to argue against something you and Guozhang left like 5 comments to request

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 3 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:1141]: The `#addNamedTopology` or `#removeNamedTopology` APIs will block until all local threads have ack'ed the topology change. I moved the un-registration to the beginning of the shutdown to prevent it from hanging if we get stuck during one of the other cleanup tasks (IIRC there is an edge case bug in the producer that can cause it to hang permanently during #close), and also to avoid blocking unnecessarily while the thread completes its shutdown, since we don't care if a removed thread gets the update or not

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java (hunk 7 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: Yep (what Guozhang said)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java (hunk 8 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java:447]: This method is used a pretty large number of times and I don't want to bloat the PR with style changes, but I totally agree that this name is confusing. I used to never remember what it was and would have to click through to the method implementation and skim the code to figure out what it was doing anytime it came up in the code. Anyways I can file a followup ticket for this

Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java:447]: Filed https://issues.apache.org/jira/browse/KAFKA-13590

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java (hunk 6 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: ok, ok, you guys convinced me like 4 comments ago :P

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java (hunk 10 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: I wanted to break up the `prepareRepartitinTopics` call into explicit steps so I could get the missing source topics returned by the middle step, but I think there's a better way to achieve what we need here so I'll clean it up

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java (hunk 5 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: Good point, I think I changed how things worked halfway through this PR and forgot to update/put some things back the way they were when the changes were no longer necessary (like this and inlining the `prepareRepartitionTopics` method)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java [hunk 11]
```diff
@@ -488,23 +492,37 @@ private boolean checkMetadataVersions(final int minReceivedMetadataVersion,
...
         );
-        repartitionTopics.setup();
-        return repartitionTopics.topicPartitionsInfo();
+        final boolean isMissingInputTopics = !repartitionTopics.setup();
+        if (isMissingInputTopics) {
+            if (!taskManager.topologyMetadata().hasNamedTopologies()) {
+                throw new MissingSourceTopicException("Missing source topics.");
+            } else {
+
+            }
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: Did you intend to leave the the else branch empty? If yes, I would simply remove it.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RepartitionTopics.java [hunk 12]
```diff
@@ -59,50 +58,92 @@ public RepartitionTopics(final TopologyMetadata topologyMetadata,
...
+            if (missingSourceTopics.isEmpty()) {
+                allRepartitionTopicConfigs.putAll(repartitionTopicConfigsPerTopology);
+                allTopicsInfo.addAll(topology.getValue());
+            } else {
+                missingUserInputTopicsPerTopology.put(topologyName, missingSourceTopics);
+                log.error("Topology {} was missing source topics {} and will be excluded from the current assignment, "
+                              + "this can be due to the consumer client's metadata being stale or because they have "
+                              + "not been created yet. Please verify that you have created all input topics. When the "
+                              + "metadata is updated a new rebalance will be kicked off automatically and the topology "
+                              + "will retried at that time.", topologyName, missingSourceTopics);
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RepartitionTopics.java]: Just commenting again here since my previous comments got outdated due to unrelated changes:
https://github.com/apache/kafka/pull/11600#discussion_r768605785
https://github.com/apache/kafka/pull/11600#discussion_r768604129

File: streams/src/test/java/org/apache/kafka/streams/integration/NamedTopologyIntegrationTest.java [hunk 13]
```diff
@@ -589,6 +580,52 @@ public void shouldAddToEmptyInitialTopologyRemoveResetOffsetsThenAddSameNamedTop
...
+        streams2.start(topology1Builder2.build());
+        waitForApplicationState(asList(streams, streams2), State.RUNNING, Duration.ofSeconds(30));
+        assertThat(waitUntilMinKeyValueRecordsReceived(consumerConfig, OUTPUT_STREAM_1, 3), equalTo(COUNT_OUTPUT_DATA));
+
+        topology2Builder.stream(NEW_STREAM).groupBy((k, v) -> k).count(IN_MEMORY_STORE).toStream().to(OUTPUT_STREAM_2);
+        topology2Builder2.stream(NEW_STREAM).groupBy((k, v) -> k).count(IN_MEMORY_STORE).toStream().to(OUTPUT_STREAM_2);
+
+        streams.addNamedTopology(topology2Builder.build());
+        streams2.addNamedTopology(topology2Builder2.build());
+
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/integration/NamedTopologyIntegrationTest.java:602]: Shouldn't you verify if topology 1 still produces output records at this point? When I read the test name I would expect that verification here.

File: streams/src/test/java/org/apache/kafka/streams/integration/NamedTopologyIntegrationTest.java [hunk 14]
```diff
@@ -589,6 +580,52 @@ public void shouldAddToEmptyInitialTopologyRemoveResetOffsetsThenAddSameNamedTop
...
+            CLUSTER.createTopic(NEW_STREAM, 2, 1);
+            produceToInputTopics(NEW_STREAM, STANDARD_INPUT_DATA);
+            assertThat(waitUntilMinKeyValueRecordsReceived(consumerConfig, OUTPUT_STREAM_2, 3), equalTo(COUNT_OUTPUT_DATA));
+        } finally {
+            CLUSTER.deleteTopicsAndWait(NEW_STREAM);
+        }
+    }
+
+    @Test
+    public void shouldWaitForMissingInputTopicsToBeCreatedWhileOtherTopologyContinuesProcessing() throws Exception {
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/integration/NamedTopologyIntegrationTest.java]: The content of this test does not fit with the name of the test. There is no other topology that continues processing.

**Review by cadonna — COMMENTED** [review]
@ableegoldman Thanks for the updates!

Here my comments.

Actually I have the feeling that I prematurely reviewed the PR since some parts seemed  WIP to me. If I did, I apologize! Maybe some comments can still be useful to you.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RepartitionTopics.java (hunk 2 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/RepartitionTopics.java]: That's what I was trying to get across, but I can try to be more clear

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java (hunk 11 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: I was working on debugging a test before pushing the changes to invoke the uncaught exception handler, I think I'm just going to split that out into a separate PR after all. I'll clean this up for this PR

File: streams/src/test/java/org/apache/kafka/streams/integration/NamedTopologyIntegrationTest.java (hunk 13 above)
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/integration/NamedTopologyIntegrationTest.java:602]: Ah good catch, added to the test

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/RepartitionTopicsTest.java [hunk 15]
```diff
@@ -162,14 +169,17 @@ public void shouldThrowMissingSourceTopicException() {
             "[test] "
         );
 
-        assertThrows(MissingSourceTopicException.class, repartitionTopics::setup);
+        assertThat(repartitionTopics.setup(), equalTo(false));
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/RepartitionTopicsTest.java:172]: Could you also verify that in the cases without missing source topics `setup()` returns `true` in the other unit tests?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java [hunk 16]
```diff
@@ -488,21 +492,32 @@ private boolean checkMetadataVersions(final int minReceivedMetadataVersion,
...
             logPrefix
         );
-        repartitionTopics.setup();
-        return repartitionTopics.topicPartitionsInfo();
+        final boolean isMissingInputTopics = !repartitionTopics.setup();
+        if (isMissingInputTopics) {
+            if (!taskManager.topologyMetadata().hasNamedTopologies()) {
+                throw new MissingSourceTopicException("Missing source topics.");
+            }
+        }
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:519]: That is just a nit that you might consider in the follow-up PR: I would not return a boolean from `setup()` since actually the setup succeeds even if the return value is `false`. I would check the return value of `RepartitionTopics#missingUserInputTopicsPerTopology()`. You could even move all this checks into `RepartitionTopics`. I think, it would make the code a bit cleaner.

**Review by cadonna — APPROVED** [review]
Thanks for the updates @ableegoldman !

LGTM!

I would just like to see the verification of the return value of `setup()` in the unit tests.

My other comment is just a proposal that you might consider in the follow-up PR or not.

**Comment by ableegoldman** [comment]
Test failures are unrelated, will address @cadonna 's remaining feedback in the followup PR https://github.com/apache/kafka/pull/11686 -- going to merge

failures:
- kafka.server.ReplicaManagerTest.[1] usesTopicIds=true
- kafka.server.LogDirFailureTest.testProduceErrorFromFailureOnLogRoll()
